compiler_opt/distributed/local/buffered_scheduler.py

```diff
@@ -0,0 +1,82 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An optimal push-pull-based load balancer which attempts to maintain at least
+`buffer` tasks assigned to each assignee.
+"""
+
+import concurrent.futures
+import threading
+
+from typing import List, Callable, TypeVar
+
+
+class DeadlockDetected(Exception):
+
+  def __init__(self):
+    Exception.__init__(self)
+
+
+def raise_deadlock(*args, **kwargs):
+  # This is raised if attempting to call .result() on _NEVER_FUTURE
+  raise DeadlockDetected()
+
+
+_NEVER_FUTURE = concurrent.futures.Future()  # A Future that never completes.
+_NEVER_FUTURE.result = raise_deadlock
+T = TypeVar('T')
+
+
+def schedule(assignments: List[Callable[[T], concurrent.futures.Future]],
+             assignees: List[T],
+             buffer=2) -> List[concurrent.futures.Future]:
+  """
+  Assigns assignments to assignees once previous assignments of the assignee are
+  completed.
+  Args:
+    assignments: Function to call with an assignee.
+    assignees: List of assignees that are the singular argument to callable.
+    buffer: Number of assignments to maintain on each assignee.
+  Returns:
+    A list of Futures that should be stored AS IS. Treat as a reference-to-list.
+  """
+  # Create "fake" futures to be returned first, these futures will be replaced
+  # with "real" futures later on.
+  results = [_NEVER_FUTURE] * len(assignments)
+  idx = -1
+  idx_lock = threading.Lock()
+
+  # Simple atomic increment and get.
+  def fetch_idx():
+    nonlocal idx
+    with idx_lock:
+      idx += 1
+      return idx
+
+  def chain_assignments(assignee):
+    if (i := fetch_idx()) < len(assignments):
+      results[i] = assignments[i](assignee)
+
+      # It is necessary for the done callback to execute on a separate thread,
```

Comment:
is this about `idx_lock`? How would it be held and not released, its release depends just on incrementing idx, and it's not waiting for anything else?

Reply:
nope- see test_deadlock. 

The actual scenario was that the callback gets called upon future.set_result(), with that callstack- this happened to be in msg_pump of the local_worker stubs, which holds the lock to the future map while doing so. The callback function also wants to fetch the future map lock so it can place a new future on it, and there's the deadlock.

Technically, we can sacrifice a bit of correctness of the scheduler (by having it not spawn the thread and do the call directly) by releasing the lock in msg_pump before doing the set_result, but it'd be a pain to debug (from my experience) if it happens.